### PR TITLE
Add Firebase authentication

### PR DESCRIPTION
**Pull request by WickyNilliams**
### What is the context of this PR?
Adds auth with firebase. Treats most routes as private, requiring auth before they can be accessed.

**You need to add your own firebase credentials to run the app. The README was updated for this PR, and explains what is needed**

### How to review 
Have a gander at code, check it works, ensure auth "flow" feels right

### CODE REVIEW

File: src/components/Header/index.js [hunk 1]
```diff
@@ -39,6 +43,30 @@ export const UtilityBtns = styled(ButtonGroup)`
   align-items: center;
 `;
 
+const UserAvatar = styled.img`
+  width: 2em;
```
Comment (hamishtaplin) [inline src/components/Header/index.js]: Can we bump this up to `2.5em`? Was a bit bigger in the mockups and makes the tooltip look a bit tidier.

**Comment by WickyNilliams** [comment]
[Integrate user sign in](https://trello.com/c/itrglzyg/296-integrate-user-sign-in)

File: src/components/UserProfile/index.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+export const LogoutButton = styled(Button)`
+  padding-left: 0;
+  padding-right: 0;
+`;
+
+const UserProfile = ({ user, onSignOut }) => (
+  <Tooltip content="Sign Out">
+    <LogoutButton clear onClick={onSignOut}>
+      <UserAvatar src={user.photoURL} alt="" />
```
[the remark this answers is not in the document]
Comment (WickyNilliams) [reply src/components/UserProfile/index.js]: Left empty alt, supplemented with `role="presentation"`, discussed

File: README.md [hunk 3]
```diff
@@ -130,6 +130,30 @@ REACT_APP_SECRET_CODE=abcdef
...
+REACT_APP_FIREBASE_PROJECT_ID=ABC REACT_APP_FIREBASE_API_KEY=DEF REACT_APP_FIREBASE_MESSAGING_SENDER_ID=GHI yarn start
+```
+
+Or they can be added to an `.env.local` file in the root of the repo:
+
+```
+REACT_APP_FIREBASE_PROJECT_ID="ABC"
+REACT_APP_FIREBASE_API_KEY="DEF"
+REACT_APP_FIREBASE_MESSAGING_SENDER_ID="GHI"
+```
```
Comment (samiwel) [inline README.md]: To test this, do I need to create my own firebase app, or is there already an eq-author app that we have the project_id and api_keys for?

Comment (samiwel) [reply README.md]: Got it working by creating own app 👍

File: src/redux/auth/reducer.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        verifiedStatus: true
+      };
+    default:
+      return state;
+  }
+};
+
+export const getUser = state => get(state, "auth.user");
+export const isSignedIn = state => Boolean(getUser(state));
+export const verifiedAuthStatus = state => get(state, "auth.verifiedStatus");
```
Comment (samiwel) [inline src/redux/auth/reducer.js]: Nice use of these utility methods 👍

**Comment by samiwel** [comment]
Probably just need to think about how we manage the firebase app used in staging. i.e. is there some ONS account we should be creating the app under, can we add the developers as admins to that app, and we need to raise a PR request to eq-compose so that the firebase project id and api keys are passed in as env vars.

**Pull request merged**